lib/rubocop/cop/lint/out_of_range_regexp_ref.rb

```diff
@@ -84,6 +90,19 @@ def on_nth_ref(node)
 
         private
 
+        def patterns(pattern_node)
+          pattern = pattern_node.node_parts[0]
+
+          case pattern.type
+          when :array_pattern, :match_alt
```

Comment:
@fatkodima @bbatsov 
I may be wrong, as I only glanced at it, but I don't think it supports hash pattern/find pattern/value pattern with the pin operator(`^(expression)`).

```ruby
{a: 'foo'} => {a: /(.)/}
p $1

['foo'] => [*, /(.)/, *]
p $1

'foo' => ^(/(.)/)
p $1
```

The last one may not need to be fixed since it is difficult to fully support and is not a practical code.